### PR TITLE
MSL: Support inline uniform blocks in argument buffers.

### PR DESCRIPTION
**Pull request by cdavis5e**
Here, the inline uniform block is explicit: we instantiate the buffer
block itself in the argument buffer, instead of a pointer to the buffer.
I just hope this will work with the `MTLArgumentDescriptor` API...

Note that Metal recursively assigns individual members of embedded
structs IDs. This means for automatic assignment that we have to
calculate the binding stride for a given buffer block. For MoltenVK,
we'll simply increment the ID by the size of the inline uniform block.
Then the later IDs will never conflict with the inline uniform block. We
can get away with this because Metal doesn't require that IDs be
contiguous, only monotonically increasing.

### CODE REVIEW

**Comment by HansKristian-Work** [comment]
Is this urgent? I'd like to look at this after the holidays.

File: spirv_msl.cpp [hunk 1]
```diff
@@ -9688,6 +9694,31 @@ void CompilerMSL::fix_up_shader_inputs_outputs()
 	});
 }
 
+// Returns the number of IDs that Metal would assign to an inline uniform block in an argument buffer.
+uint32_t CompilerMSL::get_inline_uniform_block_binding_stride(SPIRType &type)
```
Comment (HansKristian-Work) [inline spirv_msl.cpp]: Would this be necessary if we make sure to place the uniform buffer block at the highest ID? Then stride becomes irrelevant. I think relying on this layout is a little too sketchy for my liking.

File: spirv_msl.cpp [hunk 2]
```diff
@@ -9761,20 +9799,11 @@ uint32_t CompilerMSL::get_metal_resource_index(SPIRVariable &var, SPIRType::Base
...
-		desc_set = get_decoration(var.self, DecorationDescriptorSet);
-		allocate_argument_buffer_ids = descriptor_set_is_argument_buffer(desc_set);
-	}
-
 	if (allocate_argument_buffer_ids)
 	{
 		// Allocate from a flat ID binding space.
-		resource_index = next_metal_resource_ids[desc_set];
-		next_metal_resource_ids[desc_set] += binding_stride;
+		resource_index = next_metal_resource_ids[var_desc_set];
```
Comment (HansKristian-Work) [inline spirv_msl.cpp]: Why this change?

**Comment by cdavis5e** [comment]
> My main worry here is how we assign data to ids. For a pipeline layout to work in MoltenVK, we need to be able to assign a UBO blob to a single ID, irrespective of the underlying data type in the shader. If Metal requires us to assign data per sub-ID (one call for each vector) in the argument buffer, I don't think it makes sense to support this feature in SPIRV-Cross, but I don't have an Apple device handy until next year.

From the docs for `-[MTLArgumentEncoder constantDataAtIndex:]`:

> Constants declared contiguously in the Metal shading language (in an array or struct) are contiguous in memory. You can encode contiguous ranges of inlined constant data through a pointer to the first constant.

So this should be OK.

File: spirv_msl.cpp (hunk 1 above)
Comment (cdavis5e) [reply spirv_msl.cpp]: I imagine this isn't a commonly used feature. Really, I'm only adding it because support for this extension was added to MoltenVK, so now I feel like I have to support it with argument buffers. I guess this would work OK.

File: spirv_msl.cpp (hunk 2 above)
Comment (cdavis5e) [reply spirv_msl.cpp]: Isn't the `desc_set` variable redundant?

Comment (HansKristian-Work) [reply spirv_msl.cpp]: Looks like it, yes. Seems fine then.

File: spirv_msl.cpp (hunk 1 above)
Comment (HansKristian-Work) [reply spirv_msl.cpp]: Alright, I think the missing piece here is to move the (single) inline uniform block to the back of the argument buffer, and make sure it works for automatically allocated IDs as well as the add_msl_resource API.

Comment (HansKristian-Work) [reply spirv_msl.cpp]: With that change I think this method is redundant.

Comment (cdavis5e) [reply spirv_msl.cpp]: Done.

**Comment by cdavis5e** [comment]
Sorry about the long delay. I left on vacation shortly after this, and when I got back, I found myself with work elsewhere that needed to be done.

**Comment by HansKristian-Work** [comment]
No problem, thanks!
